# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSAtom.cpp
# Commit: ef77d879cb3c
# Full Hash: ef77d879cb3c5e41e6119b508be7514968ed9d90
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2019-06-15 21:47:33
# Description:
#   Bug 1555936 - Add diagnostic asserts to check the contents of the atoms table r=jandem a=abillings
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D34515
# ==============================================================================

diff -r 52f3d1a5b791 -r ef77d879cb3c js/src/vm/JSAtom.cpp
--- a/js/src/vm/JSAtom.cpp	Sat Jun 15 12:54:47 2019 +0300
+++ b/js/src/vm/JSAtom.cpp	Tue Jun 11 15:04:29 2019 +0100
@@ -398,6 +398,7 @@
 inline void AtomsTable::tracePinnedAtomsInSet(JSTracer* trc, AtomSet& atoms) {
   for (auto r = atoms.all(); !r.empty(); r.popFront()) {
     JSAtom* atom = r.front().unbarrieredGet();
+    MOZ_DIAGNOSTIC_ASSERT(atom);
     if (atom->isPinned()) {
       TraceRoot(trc, &atom, "interned_atom");
       MOZ_ASSERT(r.front().unbarrieredGet() == atom);
@@ -471,8 +472,12 @@
     AtomSet& atoms = partitions[i]->atoms;
     for (AtomSet::Enum e(atoms); !e.empty(); e.popFront()) {
       JSAtom* atom = e.front().unbarrieredGet();
+      MOZ_DIAGNOSTIC_ASSERT(atom);
       if (IsAboutToBeFinalizedUnbarriered(&atom)) {
+        MOZ_ASSERT(!atom->isPinned());
         e.removeFront();
+      } else {
+        MOZ_ASSERT(atom == e.front().unbarrieredGet());
       }
     }
   }
@@ -584,8 +589,12 @@
     }
 
     JSAtom* atom = atomsToSweep.front();
+    MOZ_DIAGNOSTIC_ASSERT(atom);
     if (IsAboutToBeFinalizedUnbarriered(&atom)) {
+      MOZ_ASSERT(!atom->isPinned());
       atomsToSweep.removeFront();
+    } else {
+      MOZ_ASSERT(atom == atomsToSweep.front());
     }
     atomsToSweep.popFront();
   }
